### PR TITLE
Mask padded video frames from attention

### LINKED ISSUE #131: Mask padded video frames from attention in the VLM video path
## What
Short or undecodable video clips are padded to `max_frames` with blank (zero) frames, and the data pipeline already produces a `frame_mask` (`True` = real frame). But that mask isn't consumed: the blank frames still become visual tokens that real tokens — and the text caption — attend to, and in MoE arches the padded tokens compete for expert capacity. This is the "frame-mask-aware attention" item deferred from the video-understanding work.
  
Consume `frame_mask` so padded-frame visual tokens are masked out of attention (and out of MoMa's expert-choice routing) across all four archs. Beyond the immediate correctness win (real-token outputs stop depending on blank-frame content), this is the foundation for variable-length / mixed image+video batches.

## Scope
- `kempnerforge/model/modality.py` — a generic per-token validity field `ModalityContext.key_padding_mask (B, S)`.
- `kempnerforge/model/vlm.py` — expand `frame_mask (B,F)` → per-visual-token `(B, F·P′)`; the four strategies place it (image-prefix arches → `key_padding_mask`, Cross-Attention → `image_mask`).
- `kempnerforge/model/{attention,mot,cross_attention}.py` — consume the mask (combine with causal; NaN guard for fully-masked rows). 
- `kempnerforge/model/moma.py` — `MoMaFFN` drops padded positions from expert-choice routing.
- `kempnerforge/model/transformer.py` — thread the mask to the blocks.
- `scripts/train.py` — pass `batch["frame_mask"]` into the forward.
- No config changes: masking is always-on and a no-op when nothing is padded.

## Backward compatibility
- Preserves current behavior wherever nothing is padded: the image (`F=1`) and text paths are unchanged (mask is `None`/all-valid → no-op), and a full clip is unchanged. 
- Pure mask — **no new state-dict keys**, so checkpoints stay compatible both directions.
- Behavior *intentionally* changes for clips with padded frames (real tokens no longer attend to blank frames) — that is the fix.

### PR DESCRIPTION
**Pull request by amazloumi**
## Summary
- Consume the per-frame `frame_mask` so real tokens never attend to padded-frame visual tokens, across all four archs. One generic per-token validity mask, `ModalityContext.key_padding_mask (B, S)`, threads through the model — consumers are modality-agnostic.
- Shared `Attention` ANDs it with the causal (and doc) mask — covers Joint-Decoder + MoMa; `MoTAttention` builds an explicit causal-AND-valid mask; Cross-Attention masks the padded image K/V via its existing`image_mask`.
- `MoMaFFN` excludes padded positions from expert-choice routing — otherwise padded tokens consume expert capacity and perturb which real tokens get processed (caught by a per-arch invariance test). 
- NaN guard unmasks fully-masked query rows (an all-padded / undecodable clip) so softmax stays finite. 
- `vlm.py`: `_visual_token_mask` expands `frame_mask (B,F)` → `(B, F·P′)`; the four strategies place it. `scripts/train.py` threads `batch["frame_mask"]`.
- **Pure mask — no new state-dict keys** (checkpoints stay compatible both ways); image (`F=1`) and text paths are unchanged (no-op when nothing is padded). Foundation for variable-length / mixed image+video batches.
- Docs + CHANGELOG updated; "frame-mask-aware attention" removed from the video Deferred list.
- **Follow-up (out of scope):** MoT with an *MoE* FFN still routes padded tokens through the shared `MoEMLP` — a "generic token-validity in MoE" change (also fixes padded text). MoT-dense (default) and MoMa are fully masked here. 
    
## Testing
- [x] `uv run ruff check kempnerforge/ tests/` passes
- [x] `uv run ruff format --check kempnerforge/ tests/ scripts/` passes
- [x] `uv run pyright kempnerforge/` passes (0 errors) 
- [x] `uv run pytest tests/unit/ -v --timeout=60` passes — affected files + the doc_ids/packing regression: 234 passed (new: `test_vlm.py::TestFramePaddingMask` — per-arch masking invariance, image no-op, undecodable-clip NaN guard, mask expansion; `test_moma.py` FFN routing exclusion; `test_modality_context.py` invariant)
- [x] If distributed code changed: `uv run torchrun --nproc_per_node=4 -m pytest tests/distributed/ -v`  ← `parallel.py` unchanged, but the model forward (attention/mot/moma) changed; worth running on GPUs 
> `[x]` e2e: **24 passed, 6 failed** — all 6 failures are **pre-existing on `main`** (PP/checkpoint/HF/sigterm infra, confirmed via baseline run) and **unrelated to this PR**; tracked in #133. The basic FSDP/TP e2e tests, which exercise the changed model forward, pass.

Closes #131

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -65,7 +65,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
...asked from attention (all four archs).** Short/undecodable video clips pad to `max_frames` with blank frames; the `frame_mask` is now consumed so real tokens never attend to padded-frame visual tokens. One per-token validity mask, `ModalityContext.key_padding_mask` `(B, S)`, threads through the model: the shared `Attention` ANDs it with the causal (and doc) mask — covering Joint-Decoder and MoMa — `MoTAttention` builds an explicit causal-AND-valid mask, and Cross-Attention masks the padded image K/V via its existing `image_mask`. A NaN guard unmasks fully-masked query rows (an all-padded clip) so softmax stays finite. It is a **pure mask — no new state-dict keys**, so checkpoints stay compatible both ways; the image (`F=1`) and text paths are unchanged (no mask is built, so they keep the FlashAttention-2 path). Note: video always takes the explicit-mask SDPA path (FA2 disabled, a `(B,1,S,S)` mask materialized) even for fully-decoded clips — the result is identical to causal-only but not free; a deliberate `torch.compile`/DP-friendly trade-off (one graph, no host sync), with FA2-recovery / FlexAttention left as a follow-up. Foundation for variable-length / mixed image+video batches.
```
[the remark this answers is not in the document]
Comment (amazloumi) [reply CHANGELOG.md]: @copilot reevaluate this and reply if it is resolved

File: docs/how-to/train-on-video.md [hunk 2]
```diff
@@ -102,9 +102,16 @@ time, so it is set in the TOML, not via a `--vlm.arch=` CLI override.)
...
-  `max_frames` with blank frames; a `frame_mask` is produced but not yet
-  consumed by the attention mask.
+- **Padded frames are masked from attention** — short/undecodable clips pad to
+  `max_frames` with blank frames, and the `frame_mask` is consumed so real
+  tokens never attend to padded-frame visual tokens (MoMa also drops them from
+  expert-choice routing); a NaN guard keeps an all-padded clip finite. It is a
+  pure mask (no new checkpoint keys); image/text keep the FlashAttention-2 path,
+  but video always takes the explicit-mask SDPA path (FA2 disabled, a `(B,1,S,S)`
+  mask built) even for fully-decoded clips — a deliberate compile/DP-friendly
+  trade-off; recovering FA2 / FlexAttention is a follow-up. *Remaining:* MoT
```
[the remark this answers is not in the document]
Comment (amazloumi) [reply docs/how-to/train-on-video.md]: @copilot reevaluate this and reply if it is resolved

File: kempnerforge/model/attention.py [hunk 3]
```diff
@@ -177,14 +183,42 @@ def forward(
...
             self.last_attention_weights = attn_weights.detach().cpu()
-        elif doc_ids is not None:
+        elif doc_ids is not None or key_padding_mask is not None:
+            # An explicit attn_mask is not a FlashAttention-2 shape, so SDPA falls
+            # back to the mem-efficient/math kernel here. Video always passes a
+            # key_padding_mask (all-True when unpadded), so video training always
+            # takes this branch -- losing FA2 and materializing a (B, 1, S, S) mask
+            # even for fully-decoded clips. Deliberate: always-masking is
+            # torch.compile / DP-friendly (one graph, no host sync). Recovering FA2
+            # for unpadded batches (or moving to FlexAttention) is a follow-up.
```
[the remark this answers is not in the document]
Comment (amazloumi) [reply kempnerforge/model/attention.py]: @copilot reevaluate this and reply if it is resolved

**Review by Naeemkh — APPROVED** [review]
